userland/examples/adc/main.c

```diff
@@ -0,0 +1,46 @@
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include "tock.h"
+#include "firestorm.h"
+#include "tock_str.h"
+#include "adc.h"
+
+int _reading = 0;
+
+static void adc_callback (
+  int callback_type __attribute__ ((unused)),
+  int channel __attribute__ ((unused)),
+  int sample,
+  void* callback_args __attribute__ ((unused))) {
+  _reading = sample;
+}
+
+int main() {
+  putstr("[Tock] ADC Test\n");
+
+  // Setup the ADC
+  adc_set_callback(adc_callback, NULL);
+  adc_initialize();
+  delay_ms(1000);
+
+  while (1) {
+    char buf[64];
+
+    // Sample channel 1. (On Firestorm, this is labeled "A5".)
+    adc_single_sample(1);
+    yield();
```

Comment:
you should always `yield_for` in these cases since you don't know which callback will fire first. For example, `putstr` doesn't actually yield (it lets the UART go asynchronously and free's the buffer whenever the callback happens to occue).